File: 백준/Silver/14501. 퇴사/퇴사.py
import sys
input = sys.stdin.readline
n = int(input())
arr = [[0,0]]
for _ in range(n):
    arr.append(list(map(int, input().split())))

dp = [0] * (n + 2)

ans = 0
for i in range(1, n + 2):
    dp[i] = max(dp[i], dp[i - 1])
    if i < n + 1:
        t, p = arr[i][0], arr[i][1]
        if i + t > n + 1:
            continue
        dp[i + t]  = max(dp[i + t], dp[i] + p)

print(max(dp))

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

class Work{
	int time;
	int money;
	Work(int time, int money){
		this.time = time;
		this.money = money;
	}
}
public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		int n = Integer.parseInt(br.readLine());
		Work[] works = new Work[n + 1];
		for (int i = 1; i <= n; i++) {
			st = new StringTokenizer(br.readLine());
			int t = Integer.parseInt(st.nextToken());
			int p = Integer.parseInt(st.nextToken());
			works[i] = new Work(t, p);
		}
		int ans = 0;
		int[] dp = new int[n + 2];

		for (int i = 1; i <= n ; i++) {
			if(i + works[i].time<= n + 1) {
				dp[i] = dp[i] + works[i].money;
				ans = Math.max(dp[i], ans);
				for (int j = i + works[i].time; j <= n + 1 ; j++) {					
					dp[j] = Math.max(dp[j], dp[i]);
				}
			}
			//System.out.println(Arrays.toString(dp));
		}
		System.out.println(ans);
	}
}
